File: src/hlhandler/wallet/manager.py
"""Wallet manager for handling private keys and addresses."""

from dataclasses import dataclass

from eth_account import Account

from hlhandler.utils import normalize_private_key, validate_private_key
from hlhandler.wallet.providers.base import KeyProvider
from hlhandler.wallet.providers.env import EnvKeyProvider
from hlhandler.wallet.providers.keyring_provider import KeyringProvider
from hlhandler.wallet.providers.prompt import PromptKeyProvider


@dataclass
class KeyResult:
    """Result of a key lookup."""

    key: str
    provider: str

    @property
    def address(self) -> str:
        """Derive Ethereum address from the private key."""
        return Account.from_key(self.key).address


class WalletManager:
    """Manages private keys using a chain of providers.

    Providers are tried in order: Env -> Keyring -> Prompt
    """

    def __init__(
        self,
        providers: list[KeyProvider] | None = None,
        allow_prompt: bool = True,
    ):
        """Initialize the wallet manager.

        Args:
            providers: Custom list of providers. If None, uses defaults.
            allow_prompt: Whether to include the prompt provider.
        """
        if providers is not None:
            self._providers = providers
        else:
            self._providers = [
                EnvKeyProvider(),
                KeyringProvider(),
            ]
            if allow_prompt:
                self._providers.append(PromptKeyProvider())

        # Keep reference to keyring provider for save/remove operations
        self._keyring: KeyringProvider | None = None
        for p in self._providers:
            if isinstance(p, KeyringProvider):
                self._keyring = p
                break

    @property
    def providers(self) -> list[KeyProvider]:
        """Get the list of providers."""
        return self._providers

    def get_private_key(self, network: str) -> KeyResult | None:
        """Get private key from the first available provider.

        Args:
            network: Network name (mainnet, testnet).

        Returns:
            KeyResult with the key and provider name, or None if not found.

        Raises:
            ValueError: If the key is invalid.
        """
        for provider in self._providers:
            if not provider.is_available():
                continue

            key = provider.get_key(network)
            if key is not None:
                # Validate and normalize the key
                if not validate_private_key(key):
                    raise ValueError(
                        f"Invalid private key from {provider.name}: "
                        "must be 64 hex characters (32 bytes)"
                    )
                normalized = normalize_private_key(key)
                return KeyResult(key=normalized, provider=provider.name)

        return None

    def get_address(self, network: str) -> str | None:
        """Get the Ethereum address for a network.

        Args:
            network: Network name (mainnet, testnet).

        Returns:
            Ethereum address or None if no key available.
        """
        result = self.get_private_key(network)
        if result:
            return result.address
        return None

    def save_to_keyring(self, network: str, key: str) -> None:
        """Save a private key to the system keyring.

        Args:
            network: Network name (mainnet, testnet).
            key: Private key to save.

        Raises:
            ValueError: If the key is invalid.
            RuntimeError: If keyring provider is not available.
        """
        if not validate_private_key(key):
            raise ValueError("Invalid private key: must be 64 hex characters (32 bytes)")

        if self._keyring is None:
            # Create a keyring provider if not in the chain
            self._keyring = KeyringProvider()

        if not self._keyring.is_available():
            raise RuntimeError("System keyring is not available")

        normalized = normalize_private_key(key)
        self._keyring.set_key(network, normalized)

    def remove_from_keyring(self, network: str) -> bool:
        """Remove a private key from the system keyring.

        Args:
            network: Network name (mainnet, testnet).

        Returns:
            True if the key was removed, False if it didn't exist.
        """
        if self._keyring is None:
            self._keyring = KeyringProvider()

        return self._keyring.delete_key(network)

    def check_providers(self, network: str) -> dict[str, dict]:
        """Check the status of all providers for a network.

        Args:
            network: Network name (mainnet, testnet).

        Returns:
            Dict with provider status information.
        """
        status = {}
        for provider in self._providers:
            available = provider.is_available()
            has_key = False

            if available:
                # Check if provider has a key without triggering prompt
                if isinstance(provider, PromptKeyProvider):
                    # Don't trigger prompt, just check cache
                    has_key = network in provider._cache
                else:
                    has_key = provider.get_key(network) is not None

            status[provider.name] = {
                "available": available,
                "has_key": has_key,
            }

        return status

File: internal/wallet/manager.go
package wallet

import (
	"errors"
	"fmt"
)

// KeyResult is a resolved private key plus the provider it came from. Mirrors
// wallet/manager.py:KeyResult.
type KeyResult struct {
	Key      string
	Provider string
}

// Address derives the Ethereum address for the key.
func (r KeyResult) Address() (string, error) { return DeriveAddress(r.Key) }

// ProviderStatus reports a provider's availability and whether it holds a key.
type ProviderStatus struct {
	Available bool
	HasKey    bool
}

// WalletManager resolves private keys through an ordered provider chain
// (env → keyring → prompt by default). Mirrors wallet/manager.py:WalletManager.
type WalletManager struct {
	providers []KeyProvider
	keyring   *KeyringProvider
}

// NewWalletManager builds the default provider chain. When allowPrompt is true
// the interactive prompt provider is appended last.
func NewWalletManager(allowPrompt bool) *WalletManager {
	kr := NewKeyringProvider()
	providers := []KeyProvider{EnvKeyProvider{}, kr}
	if allowPrompt {
		providers = append(providers, NewPromptKeyProvider())
	}
	return &WalletManager{providers: providers, keyring: kr}
}

// NewWalletManagerWithProviders builds a manager from a custom provider list.
// The first KeyringProvider in the list backs save/remove operations.
func NewWalletManagerWithProviders(providers []KeyProvider) *WalletManager {
	m := &WalletManager{providers: providers}
	for _, p := range providers {
		if kr, ok := p.(*KeyringProvider); ok {
			m.keyring = kr
			break
		}
	}
	return m
}

// Providers returns the provider chain.
func (m *WalletManager) Providers() []KeyProvider { return m.providers }

// GetPrivateKey returns the first available key from the provider chain,
// validated and normalized. Returns (nil, nil) when no provider has a key, and
// an error when a provider yields an invalid key. Mirrors
// manager.py:get_private_key.
func (m *WalletManager) GetPrivateKey(network string) (*KeyResult, error) {
	for _, provider := range m.providers {
		if !provider.IsAvailable() {
			continue
		}
		key, ok := provider.GetKey(network)
		if !ok {
			continue
		}
		if !ValidatePrivateKey(key) {
			return nil, fmt.Errorf(
				"invalid private key from %s: must be 64 hex characters (32 bytes)",
				provider.Name(),
			)
		}
		normalized, _ := NormalizePrivateKey(key)
		return &KeyResult{Key: normalized, Provider: provider.Name()}, nil
	}
	return nil, nil
}

// GetAddress returns the Ethereum address for a network, or "" when no key is
// available. Mirrors manager.py:get_address.
func (m *WalletManager) GetAddress(network string) (string, error) {
	result, err := m.GetPrivateKey(network)
	if err != nil {
		return "", err
	}
	if result == nil {
		return "", nil
	}
	return result.Address()
}

// SaveToKeyring validates and stores a key in the OS keyring. Mirrors
// manager.py:save_to_keyring.
func (m *WalletManager) SaveToKeyring(network, key string) error {
	if !ValidatePrivateKey(key) {
		return errInvalidKeyValue
	}
	if m.keyring == nil {
		m.keyring = NewKeyringProvider()
	}
	if !m.keyring.IsAvailable() {
		return errors.New("system keyring is not available")
	}
	normalized, _ := NormalizePrivateKey(key)
	return m.keyring.SetKey(network, normalized)
}

// RemoveFromKeyring removes a key from the OS keyring, returning true if it
// existed. Mirrors manager.py:remove_from_keyring.
func (m *WalletManager) RemoveFromKeyring(network string) bool {
	if m.keyring == nil {
		m.keyring = NewKeyringProvider()
	}
	return m.keyring.DeleteKey(network)
}

// CheckProviders reports each provider's availability and whether it has a key
// for the network. The prompt provider is inspected without triggering a prompt.
// Mirrors manager.py:check_providers.
func (m *WalletManager) CheckProviders(network string) map[string]ProviderStatus {
	status := make(map[string]ProviderStatus, len(m.providers))
	for _, provider := range m.providers {
		available := provider.IsAvailable()
		hasKey := false
		if available {
			if pr, ok := provider.(*PromptKeyProvider); ok {
				hasKey = pr.HasCached(network)
			} else {
				_, hasKey = provider.GetKey(network)
			}
		}
		status[provider.Name()] = ProviderStatus{Available: available, HasKey: hasKey}
	}
	return status
}

// errInvalidKeyValue mirrors the save-path Python ValueError ("Invalid private
// key: ..."), distinct from the chain-resolution message.
var errInvalidKeyValue = errors.New("invalid private key: must be 64 hex characters (32 bytes)")
